engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java

```diff
@@ -0,0 +1,1768 @@
+            val language   = EpbParser.ForeignLanguage.getBySyntacticTag(lang)
+            if (language == null) {
+              Left(s"Language $lang is not a supported polyglot language.")
+            } else {
+              val foreign = IR.Foreign.Definition(
+                language,
+                code,
+                getIdentifiedLocation(body)
+              )
+              Right((methodName, arguments, foreign))
+            }
+          case _ =>
+            Left(
+              "The body of a foreign block must be an uninterpolated string block literal."
+            )
+        }
+      case _ => Left("The method name is not specified.")
+    }
+  }
+
+  /** Translates a method reference from [[AST]] into [[IR]].
+    *
+    * @param inputAst the method reference to translate
+    * @return the [[IR]] representation of `inputAst`
+
+  def translateMethodReference(inputAst: AST): IR.Name.MethodReference = {
+    inputAst match {
+      case AstView.MethodReference(path, methodName) =>
+        val typeParts = path.map(translateExpression(_).asInstanceOf[IR.Name])
+        IR.Name.MethodReference(
+          Some(
+            IR.Name.Qualified(typeParts, MethodReference.genLocation(typeParts))
+          ),
+          translateExpression(methodName).asInstanceOf[IR.Name],
+          getIdentifiedLocation(inputAst)
+        )
+      case _ => throw new UnhandledEntity(inputAst, "translateMethodReference")
+    }
+  }
+
+  /** Translates an arbitrary program expression from {@link Tree} into {@link IR}.
+    *
+    * @param maybeParensedInput the expresion to be translated
+    * @return the {@link IR} representation of `maybeParensedInput`
+    */
+  IR.Expression translateExpression(Tree tree, boolean insideTypeSignature) {
+    return translateExpression(tree, nil(), insideTypeSignature, false);
+  }
+
+  IR.Expression translateExpression(Tree tree, List<Tree> moreArgs, boolean insideTypeSignature, boolean isMethod) {
+    return switch (tree) {
+      case null -> null;
+      case Tree.OprApp app -> {
+        var op = app.getOpr().getRight();
+        yield switch (op.codeRepr()) {
+          case "." -> {
+            var rhs = translateExpression(app.getRhs(), nil(), insideTypeSignature, true);
+            var lhs = translateExpression(app.getLhs(), insideTypeSignature);
+            IR.CallArgument callArgument = new IR$CallArgument$Specified(Option.empty(), lhs, getIdentifiedLocation(tree), meta(), diag());
+            var firstArg = cons(callArgument, nil());
+            var args = moreArgs.isEmpty() ? firstArg : translateCallArguments(moreArgs, firstArg, insideTypeSignature);
+            var prefix = new IR$Application$Prefix(
+                rhs, args,
+                false,
+                getIdentifiedLocation(tree),
+                meta(),
+                diag()
+            );
+            yield prefix;
+          }
+          default -> {
+            if (op.getBinaryInfixPrecedence() != null) {
```

Comment:
I don't think this test is necessary. If an operator doesn't have a binary infix precedence, it either is unary-only and will only appear in `UnaryOprApp` nodes, or it is unsupported, and will only appear in `Unsupported` nodes.

I plan to hide the precedence fields in Java (or remove them from the serialized format), as that information is produced by the lexer and consumed by the parser. Higher-level code shouldn't need to consider it.